### ISSUE TITLE
Hydra slaves disk full

### ISSUE THREAD

**Issue by obadz** [NixOS/nixpkgs#16995]
## Issue description

Hydra slaves' disks are often full. This causes build failures and test failures which hold back releases and waste maintainer's time.

Opening this issue so that we can better understand the problem and hopefully resolve it.

Question: what is the process that clears up disk space on Hydra slaves? Is there a periodic GC job with `max-freed = max { 0, target free space - actual free space }` ? When does it run and how much space does it attempt to free?

Eventually the problem goes away, but I don't know if that's because someone with admin rights logged into the slave and runs some GC or whether the problem clears up automatically?

What is the right long term solution to this problem? Is it a matter of getting funds for more build resources? Is there some software that needs to be written? Is there a setting that can be changed?

https://hydra.nixos.org/build/37693558/nixlog/1#line-327

cc @edolstra @vcunat @domenkozar @rbvermaa

**Comment by obadz** [comment]
cc @bjornfor

**Comment by vcunat** [comment]
AFAIK only @edolstra and @rbvermaa can do such administration. I can restart the individual jobs, for example, but I gave up during these repetitive failures. This week I also encountered out-of-disk during *evaluation*.

I agree we do need some better long-term setup. BTW, the dawin slaves didn't seem very well this week, too &ndash; we had four or five working macs on some point, and now there was one OK and one breaking builds, I think.

**Comment by grahamc** [comment]
I believe these configurations describe the GC process: https://github.com/NixOS/nixos-org-configurations/blob/master/hydra-provisioner/nixops.nix#L47-L49

Regarding funds, https://github.com/NixOS/nixos-foundation/issues/1

**Comment by edolstra** [comment]
I fixed an issue this week that was causing disk full on the central machine (https://github.com/NixOS/hydra/commit/dd32033657fc7d6a755c2feae1714148ee43fc7e), and I just bumped the GC limit from 50 GB to 100 GB. AFAIK, the main culprit here is the ability to download/view build results in the Hydra web interface, since that requires fetching store paths from the S3 binary cache.

Not sure what's up with the build slaves though. Maybe the GC frequency should be increased, but it's already running 4 times a day. I don't have access to the Macs, but it's possible they're not running the garbage collector periodically.

(Adding more machines won't help BTW until hydra-queue-runner learns about locality, i.e. becomes clever enough to assign builds to machines in a way that minimizes the number of store paths on each machine.)
